### ISSUE TITLE
Discord server

### ISSUE THREAD

**Issue by rsxdalv** [microsoft/VibeVoice#8]
Will there be a discord server for this model?

**Comment by twobob** [comment]
Here's a quick and dirty

    import discord
    from discord import app_commands
    import asyncio
    import torch
    import soundfile as sf
    import os
    import sys
    import importlib.metadata
    from typing import Optional, List

    def check_library_versions():
        """Checks for correct library versions and exits if they are wrong."""
        required_versions = {
            "transformers": "4.51.3",
            "accelerate": "1.6.0"
        }
        print("Checking required library versions...")
        all_ok = True
        for lib, req_ver in required_versions.items():
            try:
                installed_ver = importlib.metadata.version(lib)
                if installed_ver != req_ver:
                    print(f"--> Mismatch: '{lib}'. Required: {req_ver}, Installed: {installed_ver}")
                    all_ok = False
            except importlib.metadata.PackageNotFoundError:
                print(f"--> ERROR: Required library '{lib}' is not installed.")
                all_ok = False
        
        if not all_ok:
            print("\nFATAL: Incorrect library versions detected. Please fix your environment:")
            print("pip uninstall transformers accelerate")
            print("pip install transformers==4.51.3 accelerate==1.6.0")
            sys.exit(1)
        else:
            print("Library versions are correct.")

    check_library_versions()

    from transformers.utils import logging
    from VibeVoice.vibevoice.modular.modeling_vibevoice_inference import VibeVoiceForConditionalGenerationInference
    from VibeVoice.vibevoice.processor.vibevoice_processor import VibeVoiceProcessor
    from diffusers import DPMSolverMultistepScheduler

    logging.set_verbosity_error()

    DISCORD_BOT_TOKEN = os.getenv("VIBEVOICE_TOKEN", "YOUR_DISCORD_BOT_TOKEN")
    MODEL_PATH = "microsoft/VibeVoice-1.5B"
    VOICES_DIRECTORY = "voices"

    class VibeVoiceGenerator:
        def __init__(self, model_path: str, device: str = "cuda"):
            print("Loading VibeVoice model... This may take a moment.")

            if torch.cuda.is_available() and device == "cuda":
                print("CUDA is available. Loading model for GPU execution.")
                self.device = "cuda"
                model_kwargs = {
                    "torch_dtype": torch.bfloat16,
                    "device_map": "cuda",
                    "attn_implementation": "flash_attention_2",
                }
            else:
                print("Warning: CUDA not available or not selected. Loading model for CPU execution. This will be very slow.")
                self.device = "cpu"
                model_kwargs = {
                    "torch_dtype": torch.float32,
                    "device_map": "cpu",
                }

            try:
                self.processor = VibeVoiceProcessor.from_pretrained(model_path)
                self.model = VibeVoiceForConditionalGenerationInference.from_pretrained(
                    model_path,
                    **model_kwargs
                )
                self.model.eval()

                new_scheduler = DPMSolverMultistepScheduler.from_config(
                    self.model.model.noise_scheduler.config,
                    algorithm_type='sde-dpmsolver++',
                    beta_schedule='squaredcos_cap_v2'
                )
                self.model.model.noise_scheduler = new_scheduler

                self.model.set_ddpm_inference_steps(num_steps=10)
                print("VibeVoice model loaded successfully.")
            except Exception as e:
                print(f"FATAL: Failed to load model from {model_path}. Error: {e}")
                raise

        def generate_speech(self, text: str, voice_prompt_wav: str, cfg_scale: float = 1.3, output_path: str = "temp_output.wav"):
            try:
                clean_text = text.strip().replace("’", "'")
                formatted_text = clean_text
                if not formatted_text.lower().startswith('speaker '):
                    formatted_text = f"Speaker 0: {clean_text}"

                inputs = self.processor(
                    text=[formatted_text],
                    voice_samples=[voice_prompt_wav],
                    padding=True,
                    return_tensors="pt",
                    return_attention_mask=True,
                )

                outputs = self.model.generate(
                    **inputs,
                    max_new_tokens=None,
                    cfg_scale=cfg_scale,
                    tokenizer=self.processor.tokenizer,
                    generation_config={'do_sample': False},
                    verbose=False,
                )

                self.processor.save_audio(outputs.speech_outputs[0], output_path=output_path)
                return output_path
            except Exception as e:
                print(f"Error during speech generation: {e}")
                return None

    class VibeVoiceBot(discord.Client):
        def __init__(self, *, intents: discord.Intents):
            super().__init__(intents=intents)
            self.tree = app_commands.CommandTree(self)
            self.generator = VibeVoiceGenerator(model_path=MODEL_PATH)
            self.voice_presets = {}

        def load_voices(self):
            """Scans the voices directory and loads available voice presets."""
            if not os.path.isdir(VOICES_DIRECTORY):
                print(f"Warning: Voices directory '{VOICES_DIRECTORY}' not found. Please create it.")
                return

            for filename in os.listdir(VOICES_DIRECTORY):
                if filename.lower().endswith(".wav"):
                    name = os.path.splitext(filename)[0]
                    path = os.path.join(VOICES_DIRECTORY, filename)
                    self.voice_presets[name] = path
            
            if self.voice_presets:
                print(f"Loaded {len(self.voice_presets)} voices: {', '.join(self.voice_presets.keys())}")
            else:
                print("Warning: No .wav files found in the voices directory.")

        async def setup_hooks(self) -> None:
            await self.tree.sync()

    intents = discord.Intents.default()
    bot = VibeVoiceBot(intents=intents)

    @bot.event
    async def on_ready():
        print(f'Logged in as {bot.user} (ID: {bot.user.id})')
        bot.load_voices()
        print('------')

    async def voice_autocomplete(interaction: discord.Interaction, current: str) -> List[app_commands.Choice[str]]:
        """Autocomplete function for the 'voice' parameter."""
        voices = bot.voice_presets.keys()
        return [
            app_commands.Choice(name=voice, value=voice)
            for voice in voices if current.lower() in voice.lower()
        ][:25]

    @bot.tree.command(name="generate", description="Generates an audio file from text and uploads it to the channel.")
    @app_commands.describe(
        text="The text to convert to speech.",
        voice="The voice to use for the generation. Starts an autocomplete menu.",
        cfg_scale="Guidance strength. Lower is more creative, higher is more strict. Default is 1.3."
    )
    @app_commands.autocomplete(voice=voice_autocomplete)
    async def generate(interaction: discord.Interaction, text: str, voice: Optional[str] = None, cfg_scale: Optional[float] = 1.3):
        await interaction.response.defer()

        if not bot.voice_presets:
            await interaction.followup.send("Error: No voice presets are loaded. Please check the `voices` directory.", ephemeral=True)
            return

        voice_prompt_path = ""
        voice_name = ""
        if voice:
            if voice in bot.voice_presets:
                voice_prompt_path = bot.voice_presets[voice]
                voice_name = voice
            else:
                await interaction.followup.send(f"Error: Voice '{voice}' not found. Please choose one from the list.", ephemeral=True)
                return
        else:
            default_voice_name = list(bot.voice_presets.keys())[0]
            voice_prompt_path = bot.voice_presets[default_voice_name]
            voice_name = f"{default_voice_name} (Default)"

        if not (1.0 <= cfg_scale <= 2.0):
            await interaction.followup.send("Please choose a `cfg_scale` value between 1.0 and 2.0.", ephemeral=True)
            return
        
        output_filename = None
        try:
            output_filename = await asyncio.to_thread(
                bot.generator.generate_speech,
                text,
                voice_prompt_path,
                cfg_scale
            )
            if not output_filename:
                await interaction.followup.send("Sorry, something went wrong during audio generation.", ephemeral=True)
                return

            # Send the generated file to the channel
            response_message = f"🎙️ Audio generated with voice '{voice_name}': \"{text[:100]}...\" (CFG Scale: {cfg_scale})"
            await interaction.followup.send(content=response_message, file=discord.File(output_filename))

        except Exception as e:
            await interaction.followup.send(f"An error occurred during generation: {e}", ephemeral=True)
        finally:
            # Clean up the temporary audio file
            if output_filename and os.path.exists(output_filename):
                os.remove(output_filename)

    if __name__ == "__main__":
        if sys.platform == "win32":
            asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())

        if DISCORD_BOT_TOKEN == "YOUR_DISCORD_BOT_TOKEN":
            print("ERROR: Remember to replace 'YOUR_DISCORD_BOT_TOKEN' with your one from the developer portal.")
        else:
            bot.run(DISCORD_BOT_TOKEN)

something like that where this file lives above the git cloned  VibeVoice  directory and a folder called voice contains an example voice file (you can just copy move the provided examples into there or add your own)

I did amend a couple of import paths to include the VibeVoice  directory since I wanted this taxomony and they were relative.

left as an excercise to the reader how they want to handle the positioning of this script
adjust

 from VibeVoice.vibevoice.modular.modeling_vibevoice_inference import VibeVoiceForConditionalGenerationInference
 from VibeVoice.vibevoice.processor.vibevoice_processor import VibeVoiceProcessor

to your needs

VIBEVOICE_TOKEN is the environment variable that the system will be looking for the actual discord token in

24 Gb seems to be required

<img width="938" height="47" alt="Image" src="https://github.com/user-attachments/assets/76c9c1d5-60d6-480e-9258-d0d6d1d63e64" />

**Comment by fakerybakery** [comment]
[discord.gg/ZDEYTTRxWG](https://discord.gg/ZDEYTTRxWG)

**Comment by pengzhiliang** [comment]
Thanks for the interest! We don't have an official Discord server, but there's an active community at https://github.com/vibevoice-community/VibeVoice — feel free to check it out.
